DatadogInternal/Sources/Swizzling/MethodSwizzler.swift

```diff
@@ -6,127 +6,160 @@
-    public struct FoundMethod: Hashable {
-        let method: Method
-        let klass: AnyClass
+    /// Describes the current swizzled methods.
+    static var description: String {
+        methods.map { method_getName($0) }.description
+    }
 
-        fileprivate init(method: Method, klass: AnyClass) {
-            self.method = method
-            self.klass = klass
-        }
+    /// The hierarchy of swizzling per method.
+    private static var swizzlings: [Method: MethodSwizzling] = [:]
 
-        public static func == (lhs: FoundMethod, rhs: FoundMethod) -> Bool {
-            let methodParity = (lhs.method == rhs.method)
-            let classParity = (NSStringFromClass(lhs.klass) == NSStringFromClass(rhs.klass))
-            return methodParity && classParity
-        }
+    /// lock fro synchronizing `swizzlings` mutuations.
+    private static let lock = NSLock()
 
-        public func hash(into hasher: inout Hasher) {
-            let methodName = NSStringFromSelector(method_getName(method))
-            let klassName = NSStringFromClass(klass)
-            let identifier = "\(methodName)|||\(klassName)"
-            hasher.combine(identifier)
-        }
+    /// Synchronization point to access the swizzling nodes.
+    @discardableResult
+    fileprivate static func sync<T>(block: (inout [Method: MethodSwizzling]) -> T) -> T {
+        lock.lock()
+        defer { lock.unlock() }
+        return block(&swizzlings)
     }
+}
 
-    private var implementationCache: [FoundMethod: IMP] = [:]
-    var swizzledMethods: [FoundMethod] {
-        return Array(implementationCache.keys)
+/// Linked list of swizzled implementations.
+///
+/// This object hold the previous (origin) implementation of a
+/// method, the override closure reference, and a reference to its
+/// parent closure.
+private final class MethodSwizzling {
+    /// original implementation
+    let origin: IMP
+    /// type-erased override closure
+    let override: OverrideBox
+    /// parent swizzling
+    let parent: MethodSwizzling?
+
+    init(origin: IMP, override: OverrideBox, parent: MethodSwizzling? = nil) {
+        self.origin = origin
+        self.override = override
+        self.parent = parent
     }
+}
 
-    public static func findMethod(with selector: Selector, in klass: AnyClass) throws -> FoundMethod {
-        /// NOTE: RUMM-452 as we never add/remove methods/classes at runtime,
-        /// search operation doesn't have to wrapped in sync {...} although it's visible in the interface
-        var headKlass: AnyClass? = klass
-        while let someKlass = headKlass {
-            if let foundMethod = findMethod(with: selector, in: someKlass) {
-                return FoundMethod(method: foundMethod, klass: someKlass)
-            }
-            headKlass = class_getSuperclass(headKlass)
-        }
-        throw InternalError(description: "\(NSStringFromSelector(selector)) is not found in \(NSStringFromClass(klass))")
```

Comment:
I don't see any reason for doing this while the objc runtime function `class_getInstanceMethod` is doing the same thing.

Reply:
Seems like it was added to have this nicely formatted `InternalError`. I think that `class_getInstanceMethod` doesn't throw, but just returns optional.

Reply:
Not just that, previously it was using `class_copyMethodList` and looked at super-class if not found.
But the `class_getInstanceMethod` doc says specifically:
```
@note This function searches superclasses for implementations, whereas \c class_copyMethodList does not.
```

Reply:
does it make sense to use it instead and throw this exception when not found?

Reply:
Yes, I have added the `dd_class_getInstanceMethod` as part of the PR. It's at the bottom of the file 👍